### ISSUE TITLE
【Help needed!】How to render image from vertices, triangles and textures in Pytorch3D way?

### ISSUE THREAD

**Issue by imxtx** [facebookresearch/pytorch3d#1455]
Hi, I'm new to 3D things. I want to render images from vertices, triangles and textures. Verticies and colors are extracted from UV texture using uv position map and face indices. The UV texture was created by me. I want to optimize the UV texture with a target image. How should I implement it with Pytorch3D?

Here is the code for rendering:

```python3
face = resize(face, (256, 256))
[h, w, c] = face.shape
if c == 4:
    face = face[:, :, :3]
# use PRNet: 3d reconstruction -> get texture
uv_pos, vertices = self.get_vertices(face, lms)

texture = cv2.remap(
    face,
    uv_pos[:, :, :2].astype(np.float32),
    None,
    interpolation=cv2.INTER_NEAREST,
    borderMode=cv2.BORDER_CONSTANT,
    borderValue=(0),
)

texture = torch.tensor(texture).to(self.device)  # tensor, (h,w,c), (0,1)

# need to backpropagate gradients below
new_texture = self.get_masked_texture( # this is a data augmentation function
    self.mask_texture, self.mask_bimask, texture
)

new_colors = self.prn.get_colors_from_texture(new_texture)

# render
vertices = torch.tensor(vertices).to(self.device)
face_mask, new_image = self.render(vertices, new_colors, h, w)
```

The render function:

```python3
def render(self, vertices, new_colors, h, w):
    vis_colors = torch.ones((vertices.shape[0], 1)).to(self.device)
    face_mask = render_texture(
        vertices.T,
        vis_colors.T, 
        self.prn.triangles.T,
        h,
        w,
        c=1,
        device=self.device,
    )  # h,w,c
    face_mask = (face_mask > 0).int()  # to int
    new_image = render_texture(
        vertices.T, # get from texture using uv position map and face indices
        new_colors.T, # get from texture using face indices
        self.prn.triangles.T, # load from txt file
        h,
        w,
        c=3,
        device=self.device,
    )  # h,w,c
    return face_mask, new_image
```

The `render_texture` and helper function:

```python3
def render_texture(vertices, colors, triangles, h, w, c=3, device="cpu"):
    """render mesh by z buffer
    Args:
        vertices: 3 x nver
        colors: 3 x nver
        triangles: 3 x ntri
        h: height
        w: width
    """
    # initial
    image = torch.zeros((h, w, c)).to(device)

    depth_buffer = torch.zeros([h, w]) - 999999.0
    # triangle depth: approximate the depth to the average value of z in each
    # vertex(v0, v1, v2), since the vertices are closed to each other
    tri_depth = (
        vertices[2, triangles[0, :]]
        + vertices[2, triangles[1, :]]
        + vertices[2, triangles[2, :]]
    ) / 3.0
    tri_tex = (
        colors[:, triangles[0, :]]
        + colors[:, triangles[1, :]]
        + colors[:, triangles[2, :]]
    ) / 3.0

    for i in range(triangles.shape[1]):
        tri = triangles[:, i]  # 3 vertex indices

        # the inner bounding box
        umin = max(int(torch.ceil(torch.min(vertices[0, tri]))), 0)
        umax = min(int(torch.floor(torch.max(vertices[0, tri]))), w - 1)

        vmin = max(int(torch.ceil(torch.min(vertices[1, tri]))), 0)
        vmax = min(int(torch.floor(torch.max(vertices[1, tri]))), h - 1)

        if umax < umin or vmax < vmin:
            continue

        for u in range(umin, umax + 1):
            for v in range(vmin, vmax + 1):
                if tri_depth[i] > depth_buffer[v, u] and isPointInTriangle(
                    [u, v], vertices[:2, tri].cpu().numpy()
                ):
                    depth_buffer[v, u] = tri_depth[i]
                    image[v, u, :] = tri_tex[:, i]
    return image

def isPointInTriangle(point, tri_points):
    tp = tri_points

    # vectors
    v0 = tp[:, 2] - tp[:, 0]
    v1 = tp[:, 1] - tp[:, 0]
    v2 = point - tp[:, 0]

    # dot products
    dot00 = np.matmul(v0.T, v0)
    dot01 = np.matmul(v0.T, v1)
    dot02 = np.matmul(v0.T, v2)
    dot11 = np.matmul(v1.T, v1)
    dot12 = np.matmul(v1.T, v2)

    # barycentric coordinates
    if dot00 * dot11 - dot01 * dot01 == 0:
        inverDeno = 0
    else:
        inverDeno = 1 / (dot00 * dot11 - dot01 * dot01)

    u = (dot11 * dot02 - dot01 * dot12) * inverDeno
    v = (dot00 * dot12 - dot01 * dot02) * inverDeno

    # check if point in triangle
    return (u >= 0) & (v >= 0) & (u + v < 1)
```

Here is the code for getting vertices and colors:

```python3
def get_vertices(self, pos):
    all_vertices = np.reshape(pos, [self.resolution**2, -1])
    vertices = all_vertices[self.face_ind, :]
    return vertices

def get_colors_from_texture(self, texture):
    all_colors = torch.reshape(texture, [self.resolution**2, -1])
    colors = all_colors[self.face_ind, :]
    return colors
```

The [face indices](https://github.com/yfeng95/PRNet/blob/master/Data/uv-data/face_ind.txt) file is like this:

```text
1.285000000000000000e+03
1.286000000000000000e+03
1.287000000000000000e+03
1.288000000000000000e+03
1.289000000000000000e+03
1.290000000000000000e+03
1.291000000000000000e+03
1.292000000000000000e+03
...
```

The [triangles](https://raw.githubusercontent.com/yfeng95/PRNet/master/Data/uv-data/triangles.txt) file is like this:

```text
1.000000000000000000e+00 2.470000000000000000e+02 2.460000000000000000e+02
1.000000000000000000e+00 2.000000000000000000e+00 2.470000000000000000e+02
2.000000000000000000e+00 2.480000000000000000e+02 2.470000000000000000e+02
2.000000000000000000e+00 3.000000000000000000e+00 2.480000000000000000e+02
3.000000000000000000e+00 2.490000000000000000e+02 2.480000000000000000e+02
3.000000000000000000e+00 4.000000000000000000e+00 2.490000000000000000e+02
4.000000000000000000e+00 2.500000000000000000e+02 2.490000000000000000e+02
4.000000000000000000e+00 5.000000000000000000e+00 2.500000000000000000e+02
5.000000000000000000e+00 2.510000000000000000e+02 2.500000000000000000e+02
...
```

**Comment by imxtx** [comment]
I have texture image and verticies, but I don't know how to generate `faces_uvs` and `verts_uvs`.

```python
def render_pytorch3d(self, texture_image, faces_uvs, verts, verts_uvs):
    tex = TexturesUV(maps=[texture_image], faces_uvs=faces_uvs, verts_uvs=verts_uvs)
    # verts: (n_ver x 3)，triangles: (n_face x 3)
    mesh = Meshes(verts=[verts], faces=[self.prn.triangles], textures=tex)
    return self.renderer(mesh)
```

I have uv position map generate from PRNet. How can I generate faces_uvs and verts_uvs from it?

**Comment by bottler** [comment]
I don't really understand. When you say you "have" a texture image and vertices, what information do you know, and what do you need to find. Do you have enough information that for each triangle (given as a triple of vertices) you can give its location as a triangle in the image? If so, you just need to populate that in verts_uv and faces_uv. I.e. put all the 2D coordinates corners of triangles in the image into verts_uv and then make faces_uv have the same shape as faces (i.e. 3 integer indexes for each triangle), where each value is the index into verts_uv. 

If you don't have that much information, then maybe you want to write a model to optimize to find it? We can't help with such things.

**Comment by imxtx** [comment]
> I don't really understand. When you say you "have" a texture image and vertices, what information do you know, and what do you need to find. Do you have enough information that for each triangle (given as a triple of vertices) you can give its location as a triangle in the image? If so, you just need to populate that in verts_uv and faces_uv. I.e. put all the 2D coordinates corners of triangles in the image into verts_uv and then make faces_uv have the same shape as faces (i.e. 3 integer indexes for each triangle), where each value is the index into verts_uv.
> 
> If you don't have that much information, then maybe you want to write a model to optimize to find it? We can't help with such things.

Thank you for your reply. Texture image is a UV texture image like this:

![texture](https://user-images.githubusercontent.com/93981430/221735202-0b19dd91-bafd-4eb5-9a8e-0c12bb5f4c37.png)

Vertices are all the vertices on this UV texture image. I compute verts_uv like this:

```python3
def generate_uv_coords(self):
    resolution = self.resolution
    uv_coords = np.meshgrid(range(resolution), range(resolution))
    uv_coords = np.transpose(np.array(uv_coords), [1, 2, 0])
    uv_coords = np.reshape(uv_coords, [resolution**2, -1])
    uv_coords = uv_coords[self.face_ind, :]
    uv_coords = np.hstack((uv_coords[:, :2], np.zeros([uv_coords.shape[0], 1])))
    return (uv_coords / resolution).astype(np.float32)
```

Are faces_uv and this [triangles]() file the same thing? Or how can I generate faces_uv?

I saved the texture image created by this line of code, the result seems incorrect:

```python
meshes = Meshes(verts=[verts], faces=[self.prn.triangles_tensor], textures=tex)`:
```

```python
plt.figure(figsize=(7,7))
texture_image=meshes.textures.maps_padded()
plt.imshow(texture_image.squeeze().detach().cpu().numpy())
plt.savefig('filename')
```

<img width="650" alt="image" src="https://user-images.githubusercontent.com/93981430/221737665-aa016de2-3079-4007-b042-ec5e422aca76.png">


Here is the code for creating renderer, I followed the cow tutorial:

```python
R, T = look_at_view_transform(2.7, 0, 180)
cameras = FoVPerspectiveCameras(device=self.device, R=R, T=T)
raster_settings = RasterizationSettings(
    image_size=256,
    blur_radius=0.0,
    faces_per_pixel=1,
)
lights = PointLights(device=self.device, location=[[0.0, 0.0, -3.0]])
self.renderer = MeshRenderer(
    rasterizer=MeshRasterizer(cameras=cameras, raster_settings=raster_settings),
    shader=SoftPhongShader(device=self.device, cameras=cameras, lights=lights),
)
```

The rendered image is full of white pixels:

![new_image](https://user-images.githubusercontent.com/93981430/221737237-d449ec27-310c-474d-a1e4-e621f1ef3d37.png)

I use the plot api to visualize, but there is no camera. The shape is also strange, it should be a human face:

```python
fig = plot_batch_individually([meshes[0], self.renderer.shader.cameras])
```

![fig](https://user-images.githubusercontent.com/93981430/221736831-13a0d0e7-3194-4e88-b679-844478707548.png)

**Comment by bottler** [comment]
I can't look at the details or the data here, but I can give you some pointers: Can you see the camera object on the plotly plot? Is it where you expect it to be relative to the mesh? (This can help cases of the rendered image being unexpected.)

Your code for adding the face mask may just have been not what you meant. I don't think that's a pytorch3d thing. The texture map is just a tensor.

It can make sense for faces_uv to be the same as the mesh faces (i.e. triangles I think). That is the case if every vertex of the mesh corresponds to only one location in the texture map, e.g. where the texture is locally like a plane.

Is what you are trying to achieve something like making a mesh of a view of an object starting from an image and some notion of depth in the image?

**Comment by imxtx** [comment]
Thanks, I can't see the camera object. But now the plotly plot seems like a human face viewed from top right. But the rendered image is still totally white. Maybe I need to adjust the position of the camera？

![image](https://user-images.githubusercontent.com/93981430/221823482-19ac8fba-ad28-4805-8430-738150b9aaa4.png)

What I'm trying to do is optimizing the facial mask texture to a specific style. The render process shown bellow is very slow. I wrote a very slow renderer composed of three levels of loop. I need iterate the render process 100 times on a single image to get the style-transfered facial mask.  So I want to use pytorch3d to utilized the GPU acceleration. 

<img width="1290" alt="image" src="https://user-images.githubusercontent.com/93981430/221825166-53a00781-4424-4314-9c51-42e8539c0a3d.png">

**Comment by imxtx** [comment]
I managed to create the plotly plot. There is a red dot behind the mesh. I don't know what it is.

<img width="406" alt="image" src="https://user-images.githubusercontent.com/93981430/221856201-a154fec0-bfcf-479d-92cd-59a095cd6c72.png">

After click `reset camera` it looks like this:

<img width="1053" alt="image" src="https://user-images.githubusercontent.com/93981430/221856707-518bfadd-c3bc-403a-9b29-20e7b4111554.png">

**Comment by bottler** [comment]
The red dot (which I can't see btw) will be the camera. If you zoom in on it you should be able to see its orientation. If it isn't where you want it to be, relative to the face, then you can fix it by fixing your definition of the camera.

**Comment by imxtx** [comment]
Many thanks to you!  You helped a lot. It's the camera.

<img width="591" alt="image" src="https://user-images.githubusercontent.com/93981430/221865824-8243a7ee-f543-4b82-9444-0a5774bae9e7.png">

<img width="379" alt="image" src="https://user-images.githubusercontent.com/93981430/221867280-e5fbb90e-e9e1-4116-89bd-45a8b27c940d.png">

Could you please spare me a little time for the last question. If I want to get a rendered image with the save size as the input UV texture. How should I set the parameters of the cameras? I still get a white image.

**Comment by bottler** [comment]
For something like this, you need to work out the position of the camera yourself.

**Comment by imxtx** [comment]
Thank you for your greate help.
